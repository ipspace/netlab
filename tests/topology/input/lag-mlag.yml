#
# Advanced dual LAG MLAG example - 2x2 links between hosts and MLAG pair
#

provider: clab

module: [ lag, vlan ]

vlans:
 red:

groups:
 _auto_create: True
 switches:
  members: [s1,s2]
  device: dellos10
 hosts:
  members: [h1,h2]
  device: frr

links:
- lag: # MLAG peer link between S1 and S2, redundant
   members: [s1-s2,s2-s1]
   mlag.peergroup: True
- lag:
   members: [ h1-s1,h1-s1,h1-s2,h1-s2 ]  # Test multiple links between same pair of nodes
  vlan.access: red
- lag:
   members: [ h1-s1, h1-s1 ]             # Regression: MLAG followed by regular lag should not pick overlapping lag.ifindex
  pool: l2only
- lag:
   members: [ h2-s1,h2-s1,h2-s2,h2-s2 ]
  vlan.access: red
- lag:
   members: [ h2-s1,h2-s2 ]              # A second lag should get a unique ifindex, not overlapping with the first one
  vlan.access: red