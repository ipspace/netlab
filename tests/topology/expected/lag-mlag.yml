groups:
  hosts:
    device: frr
    members:
    - h1
    - h2
  switches:
    device: dellos10
    members:
    - s1
    - s2
input:
- topology/input/lag-mlag.yml
- package:topology-defaults.yml
lag:
  lacp: fast
  lacp_mode: active
links:
- _linkname: links[1].peerlink[1]
  interfaces:
  - ifindex: 1
    ifname: ethernet1/1/1
    node: s1
  - ifindex: 1
    ifname: ethernet1/1/1
    node: s2
  lag:
    mlag:
      peergroup: 1
  linkindex: 1
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[1].peerlink[2]
  interfaces:
  - ifindex: 2
    ifname: ethernet1/1/2
    node: s2
  - ifindex: 2
    ifname: ethernet1/1/2
    node: s1
  lag:
    _peerlink: 1
  linkindex: 6
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[2].lag[1]
  interfaces:
  - ifindex: 1
    ifname: eth1
    node: h1
  - ifindex: 3
    ifname: ethernet1/1/3
    node: s1
  lag:
    _parentindex: 2
  linkindex: 7
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[2].lag[2]
  interfaces:
  - ifindex: 2
    ifname: eth2
    node: h1
  - ifindex: 4
    ifname: ethernet1/1/4
    node: s1
  lag:
    _parentindex: 2
  linkindex: 8
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[2].lag[3]
  interfaces:
  - ifindex: 3
    ifname: eth3
    node: h1
  - ifindex: 3
    ifname: ethernet1/1/3
    node: s2
  lag:
    _parentindex: 2
  linkindex: 9
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[2].lag[4]
  interfaces:
  - ifindex: 4
    ifname: eth4
    node: h1
  - ifindex: 4
    ifname: ethernet1/1/4
    node: s2
  lag:
    _parentindex: 2
  linkindex: 10
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[3].lag[1]
  interfaces:
  - ifindex: 5
    ifname: eth5
    node: h1
  - ifindex: 5
    ifname: ethernet1/1/5
    node: s1
  lag:
    _parentindex: 3
  linkindex: 11
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[3].lag[2]
  interfaces:
  - ifindex: 6
    ifname: eth6
    node: h1
  - ifindex: 6
    ifname: ethernet1/1/6
    node: s1
  lag:
    _parentindex: 3
  linkindex: 12
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[4].lag[1]
  interfaces:
  - ifindex: 1
    ifname: eth1
    node: h2
  - ifindex: 7
    ifname: ethernet1/1/7
    node: s1
  lag:
    _parentindex: 4
  linkindex: 13
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[4].lag[2]
  interfaces:
  - ifindex: 2
    ifname: eth2
    node: h2
  - ifindex: 8
    ifname: ethernet1/1/8
    node: s1
  lag:
    _parentindex: 4
  linkindex: 14
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[4].lag[3]
  interfaces:
  - ifindex: 3
    ifname: eth3
    node: h2
  - ifindex: 5
    ifname: ethernet1/1/5
    node: s2
  lag:
    _parentindex: 4
  linkindex: 15
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[4].lag[4]
  interfaces:
  - ifindex: 4
    ifname: eth4
    node: h2
  - ifindex: 6
    ifname: ethernet1/1/6
    node: s2
  lag:
    _parentindex: 4
  linkindex: 16
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[5].lag[1]
  interfaces:
  - ifindex: 5
    ifname: eth5
    node: h2
  - ifindex: 9
    ifname: ethernet1/1/9
    node: s1
  lag:
    _parentindex: 5
  linkindex: 17
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[5].lag[2]
  interfaces:
  - ifindex: 6
    ifname: eth6
    node: h2
  - ifindex: 7
    ifname: ethernet1/1/7
    node: s2
  lag:
    _parentindex: 5
  linkindex: 18
  node_count: 2
  prefix: false
  type: p2p
module:
- lag
- vlan
name: input
nodes:
  h1:
    af:
      ipv4: true
    box: quay.io/frrouting/frr:10.0.1
    clab:
      binds:
      - clab_files/h1/daemons:/etc/frr/daemons
      - clab_files/h1/hosts:/etc/hosts
      config_templates:
      - daemons:/etc/frr/daemons
      - hosts:/etc/hosts
      kind: linux
    device: frr
    hostname: clab-input-h1
    id: 3
    interfaces:
    - ifindex: 30000
      ifname: bond1
      lag:
        ifindex: 1
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] h1 -> [s1,s2]'
      neighbors:
      - ifname: port-channel1
        ipv4: 172.16.0.1/24
        node: s1
      - ifname: port-channel1
        ipv4: 172.16.0.2/24
        node: s2
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 30001
      ifname: bond2
      lag:
        ifindex: 2
        lacp: fast
        lacp_mode: active
      name: h1 -> s1
      neighbors:
      - ifname: port-channel2
        node: s1
      pool: l2only
      type: lag
      virtual_interface: true
    - ifindex: 1
      ifname: eth1
      lag:
        _parentindex: 1
      linkindex: 7
      mtu: 1500
      name: h1 -> s1
      neighbors:
      - ifname: ethernet1/1/3
        node: s1
      type: p2p
    - ifindex: 2
      ifname: eth2
      lag:
        _parentindex: 1
      linkindex: 8
      mtu: 1500
      name: h1 -> s1
      neighbors:
      - ifname: ethernet1/1/4
        node: s1
      type: p2p
    - ifindex: 3
      ifname: eth3
      lag:
        _parentindex: 1
      linkindex: 9
      mtu: 1500
      name: h1 -> s2
      neighbors:
      - ifname: ethernet1/1/3
        node: s2
      type: p2p
    - ifindex: 4
      ifname: eth4
      lag:
        _parentindex: 1
      linkindex: 10
      mtu: 1500
      name: h1 -> s2
      neighbors:
      - ifname: ethernet1/1/4
        node: s2
      type: p2p
    - ifindex: 5
      ifname: eth5
      lag:
        _parentindex: 2
      linkindex: 11
      mtu: 1500
      name: h1 -> s1
      neighbors:
      - ifname: ethernet1/1/5
        node: s1
      type: p2p
    - ifindex: 6
      ifname: eth6
      lag:
        _parentindex: 2
      linkindex: 12
      mtu: 1500
      name: h1 -> s1
      neighbors:
      - ifname: ethernet1/1/6
        node: s1
      type: p2p
    - bridge_group: 1
      ifindex: 40000
      ifname: vlan1000
      ipv4: 172.16.0.3/24
      name: VLAN red (1000) -> [s2,s1,h2]
      neighbors:
      - ifname: virtual-network1000
        ipv4: 172.16.0.2/24
        node: s2
      - ifname: virtual-network1000
        ipv4: 172.16.0.1/24
        node: s1
      - ifname: vlan1000
        ipv4: 172.16.0.4/24
        node: h2
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.3/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.103
      mac: 08:4f:a9:00:00:03
    module:
    - lag
    - vlan
    mtu: 1500
    name: h1
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
  h2:
    af:
      ipv4: true
    box: quay.io/frrouting/frr:10.0.1
    clab:
      binds:
      - clab_files/h2/daemons:/etc/frr/daemons
      - clab_files/h2/hosts:/etc/hosts
      config_templates:
      - daemons:/etc/frr/daemons
      - hosts:/etc/hosts
      kind: linux
    device: frr
    hostname: clab-input-h2
    id: 4
    interfaces:
    - ifindex: 30000
      ifname: bond1
      lag:
        ifindex: 1
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] h2 -> [s1,s2]'
      neighbors:
      - ifname: port-channel3
        ipv4: 172.16.0.1/24
        node: s1
      - ifname: port-channel3
        ipv4: 172.16.0.2/24
        node: s2
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 30001
      ifname: bond2
      lag:
        ifindex: 2
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] h2 -> [s1,s2]'
      neighbors:
      - ifname: port-channel4
        ipv4: 172.16.0.1/24
        node: s1
      - ifname: port-channel4
        ipv4: 172.16.0.2/24
        node: s2
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 1
      ifname: eth1
      lag:
        _parentindex: 1
      linkindex: 13
      mtu: 1500
      name: h2 -> s1
      neighbors:
      - ifname: ethernet1/1/7
        node: s1
      type: p2p
    - ifindex: 2
      ifname: eth2
      lag:
        _parentindex: 1
      linkindex: 14
      mtu: 1500
      name: h2 -> s1
      neighbors:
      - ifname: ethernet1/1/8
        node: s1
      type: p2p
    - ifindex: 3
      ifname: eth3
      lag:
        _parentindex: 1
      linkindex: 15
      mtu: 1500
      name: h2 -> s2
      neighbors:
      - ifname: ethernet1/1/5
        node: s2
      type: p2p
    - ifindex: 4
      ifname: eth4
      lag:
        _parentindex: 1
      linkindex: 16
      mtu: 1500
      name: h2 -> s2
      neighbors:
      - ifname: ethernet1/1/6
        node: s2
      type: p2p
    - ifindex: 5
      ifname: eth5
      lag:
        _parentindex: 2
      linkindex: 17
      mtu: 1500
      name: h2 -> s1
      neighbors:
      - ifname: ethernet1/1/9
        node: s1
      type: p2p
    - ifindex: 6
      ifname: eth6
      lag:
        _parentindex: 2
      linkindex: 18
      mtu: 1500
      name: h2 -> s2
      neighbors:
      - ifname: ethernet1/1/7
        node: s2
      type: p2p
    - bridge_group: 1
      ifindex: 40000
      ifname: vlan1000
      ipv4: 172.16.0.4/24
      name: VLAN red (1000) -> [h1,s2,s1]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.3/24
        node: h1
      - ifname: virtual-network1000
        ipv4: 172.16.0.2/24
        node: s2
      - ifname: virtual-network1000
        ipv4: 172.16.0.1/24
        node: s1
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.4/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.104
      mac: 08:4f:a9:00:00:04
    module:
    - lag
    - vlan
    mtu: 1500
    name: h2
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
  s1:
    af:
      ipv4: true
    box: vrnetlab/vr-ftosv
    clab:
      kind: vr-ftosv
    device: dellos10
    hostname: clab-input-s1
    id: 1
    interfaces:
    - clab:
        name: eth1
      ifindex: 1
      ifname: ethernet1/1/1
      lag:
        mlag:
          mac: 02-00-01-A9-00-01
          peer: 10.0.0.2
          peergroup: 1
      linkindex: 1
      mtu: 1500
      name: s1 -> s2
      neighbors:
      - ifname: ethernet1/1/1
        node: s2
      type: p2p
    - ifindex: 30000
      ifname: port-channel1
      lag:
        _mlag: true
        ifindex: 1
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] s1 -> [h1,s2]'
      neighbors:
      - ifname: bond1
        ipv4: 172.16.0.3/24
        node: h1
      - ifname: port-channel1
        ipv4: 172.16.0.2/24
        node: s2
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 30001
      ifname: port-channel2
      lag:
        ifindex: 2
        lacp: fast
        lacp_mode: active
      name: s1 -> h1
      neighbors:
      - ifname: bond2
        node: h1
      pool: l2only
      type: lag
      virtual_interface: true
    - ifindex: 30002
      ifname: port-channel3
      lag:
        _mlag: true
        ifindex: 3
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] s1 -> [h2,s2]'
      neighbors:
      - ifname: bond1
        ipv4: 172.16.0.4/24
        node: h2
      - ifname: port-channel3
        ipv4: 172.16.0.2/24
        node: s2
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 30003
      ifname: port-channel4
      lag:
        _mlag: true
        ifindex: 4
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] s1 -> [h2,s2]'
      neighbors:
      - ifname: bond2
        ipv4: 172.16.0.4/24
        node: h2
      - ifname: port-channel4
        ipv4: 172.16.0.2/24
        node: s2
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - clab:
        name: eth2
      ifindex: 2
      ifname: ethernet1/1/2
      lag:
        _peerlink: 1
      linkindex: 6
      mtu: 1500
      name: s1 -> s2
      neighbors:
      - ifname: ethernet1/1/2
        node: s2
      type: p2p
    - clab:
        name: eth3
      ifindex: 3
      ifname: ethernet1/1/3
      lag:
        _parentindex: 1
      linkindex: 7
      mtu: 1500
      name: s1 -> h1
      neighbors:
      - ifname: eth1
        node: h1
      type: p2p
    - clab:
        name: eth4
      ifindex: 4
      ifname: ethernet1/1/4
      lag:
        _parentindex: 1
      linkindex: 8
      mtu: 1500
      name: s1 -> h1
      neighbors:
      - ifname: eth2
        node: h1
      type: p2p
    - clab:
        name: eth5
      ifindex: 5
      ifname: ethernet1/1/5
      lag:
        _parentindex: 2
      linkindex: 11
      mtu: 1500
      name: s1 -> h1
      neighbors:
      - ifname: eth5
        node: h1
      type: p2p
    - clab:
        name: eth6
      ifindex: 6
      ifname: ethernet1/1/6
      lag:
        _parentindex: 2
      linkindex: 12
      mtu: 1500
      name: s1 -> h1
      neighbors:
      - ifname: eth6
        node: h1
      type: p2p
    - clab:
        name: eth7
      ifindex: 7
      ifname: ethernet1/1/7
      lag:
        _parentindex: 3
      linkindex: 13
      mtu: 1500
      name: s1 -> h2
      neighbors:
      - ifname: eth1
        node: h2
      type: p2p
    - clab:
        name: eth8
      ifindex: 8
      ifname: ethernet1/1/8
      lag:
        _parentindex: 3
      linkindex: 14
      mtu: 1500
      name: s1 -> h2
      neighbors:
      - ifname: eth2
        node: h2
      type: p2p
    - clab:
        name: eth9
      ifindex: 9
      ifname: ethernet1/1/9
      lag:
        _parentindex: 4
      linkindex: 17
      mtu: 1500
      name: s1 -> h2
      neighbors:
      - ifname: eth5
        node: h2
      type: p2p
    - bridge_group: 1
      ifindex: 40000
      ifname: virtual-network1000
      ipv4: 172.16.0.1/24
      name: VLAN red (1000) -> [h1,s2,h2]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.3/24
        node: h1
      - ifname: virtual-network1000
        ipv4: 172.16.0.2/24
        node: s2
      - ifname: vlan1000
        ipv4: 172.16.0.4/24
        node: h2
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
    loopback:
      ifindex: 0
      ifname: loopback0
      ipv4: 10.0.0.1/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: mgmt1/1/1
      ipv4: 192.168.121.101
      mac: 08:4f:a9:00:00:01
    module:
    - lag
    - vlan
    mtu: 1500
    name: s1
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
  s2:
    af:
      ipv4: true
    box: vrnetlab/vr-ftosv
    clab:
      kind: vr-ftosv
    device: dellos10
    hostname: clab-input-s2
    id: 2
    interfaces:
    - clab:
        name: eth1
      ifindex: 1
      ifname: ethernet1/1/1
      lag:
        mlag:
          mac: 02-00-01-A9-00-01
          peer: 10.0.0.1
          peergroup: 1
      linkindex: 1
      mtu: 1500
      name: s2 -> s1
      neighbors:
      - ifname: ethernet1/1/1
        node: s1
      type: p2p
    - ifindex: 30000
      ifname: port-channel1
      lag:
        _mlag: true
        ifindex: 1
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] s2 -> [h1,s1]'
      neighbors:
      - ifname: bond1
        ipv4: 172.16.0.3/24
        node: h1
      - ifname: port-channel1
        ipv4: 172.16.0.1/24
        node: s1
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 30001
      ifname: port-channel3
      lag:
        _mlag: true
        ifindex: 3
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] s2 -> [h2,s1]'
      neighbors:
      - ifname: bond1
        ipv4: 172.16.0.4/24
        node: h2
      - ifname: port-channel3
        ipv4: 172.16.0.1/24
        node: s1
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 30002
      ifname: port-channel4
      lag:
        _mlag: true
        ifindex: 4
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] s2 -> [h2,s1]'
      neighbors:
      - ifname: bond2
        ipv4: 172.16.0.4/24
        node: h2
      - ifname: port-channel4
        ipv4: 172.16.0.1/24
        node: s1
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - clab:
        name: eth2
      ifindex: 2
      ifname: ethernet1/1/2
      lag:
        _peerlink: 1
      linkindex: 6
      mtu: 1500
      name: s2 -> s1
      neighbors:
      - ifname: ethernet1/1/2
        node: s1
      type: p2p
    - clab:
        name: eth3
      ifindex: 3
      ifname: ethernet1/1/3
      lag:
        _parentindex: 1
      linkindex: 9
      mtu: 1500
      name: s2 -> h1
      neighbors:
      - ifname: eth3
        node: h1
      type: p2p
    - clab:
        name: eth4
      ifindex: 4
      ifname: ethernet1/1/4
      lag:
        _parentindex: 1
      linkindex: 10
      mtu: 1500
      name: s2 -> h1
      neighbors:
      - ifname: eth4
        node: h1
      type: p2p
    - clab:
        name: eth5
      ifindex: 5
      ifname: ethernet1/1/5
      lag:
        _parentindex: 3
      linkindex: 15
      mtu: 1500
      name: s2 -> h2
      neighbors:
      - ifname: eth3
        node: h2
      type: p2p
    - clab:
        name: eth6
      ifindex: 6
      ifname: ethernet1/1/6
      lag:
        _parentindex: 3
      linkindex: 16
      mtu: 1500
      name: s2 -> h2
      neighbors:
      - ifname: eth4
        node: h2
      type: p2p
    - clab:
        name: eth7
      ifindex: 7
      ifname: ethernet1/1/7
      lag:
        _parentindex: 4
      linkindex: 18
      mtu: 1500
      name: s2 -> h2
      neighbors:
      - ifname: eth6
        node: h2
      type: p2p
    - bridge_group: 1
      ifindex: 40000
      ifname: virtual-network1000
      ipv4: 172.16.0.2/24
      name: VLAN red (1000) -> [h1,s1,h2]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.3/24
        node: h1
      - ifname: virtual-network1000
        ipv4: 172.16.0.1/24
        node: s1
      - ifname: vlan1000
        ipv4: 172.16.0.4/24
        node: h2
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
    loopback:
      ifindex: 0
      ifname: loopback0
      ipv4: 10.0.0.2/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: mgmt1/1/1
      ipv4: 192.168.121.102
      mac: 08:4f:a9:00:00:02
    module:
    - lag
    - vlan
    mtu: 1500
    name: s2
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
provider: clab
vlans:
  red:
    host_count: 0
    id: 1000
    neighbors:
    - ifname: vlan1000
      ipv4: 172.16.0.3/24
      node: h1
    - ifname: virtual-network1000
      ipv4: 172.16.0.2/24
      node: s2
    - ifname: virtual-network1000
      ipv4: 172.16.0.1/24
      node: s1
    - ifname: vlan1000
      ipv4: 172.16.0.4/24
      node: h2
    prefix:
      allocation: id_based
      ipv4: 172.16.0.0/24
